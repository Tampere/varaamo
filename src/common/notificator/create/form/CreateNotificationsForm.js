import React, { useRef, useState } from 'react';
import firebase from 'firebase/app';
import 'firebase/auth';
import Select from 'react-select';
import { Row, Col } from 'react-bootstrap';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { createSelectOptions } from '../../utils';
import NotificationDatePicker from '../../date/NotificatorDatePicker';

const CreateNotificationsForm = (props) => {
  const formContainer = useRef();
  const [activeLanguage, setActiveLanguage] = useState('fi');

  const {
    addElement, addNew, isEditing, newNotification, onFieldChange,
  } = props;

  const { targetOptions, urgencyOptions, activeOptions } = createSelectOptions();
  const languageButtons = ['fi', 'en', 'sv'];
  // Because this component is used inside modal, count dimensions and set sm value based on that
  const current = formContainer.current;
  let dimensions = {
    width: 601,
  };
  if (current) {
    dimensions = current.getBoundingClientRect();
  }

  return (
    <div ref={formContainer}>
      {!isEditing && (
        <React.Fragment>
          <div className="button-row">
            <button onClick={() => firebase.auth().signOut()} type="button">Sign out</button>
          </div>
          <h4>Create new notification</h4>
        </React.Fragment>
      )}
      <Row>
        <Col sm={dimensions && dimensions.width > 600 ? 3 : 12}>
          <span>Name</span>
          <input
            onChange={event => onFieldChange(event, 'name')}
            value={newNotification.name || ''}
          />
        </Col>
        <Col sm={dimensions && dimensions.width > 600 ? 3 : 12}>
          <span>Target</span>
          <Select
            className="app-Select"
            classNamePrefix="app-Select"
            onChange={event => onFieldChange({ target: { value: event } }, 'target')}
            options={targetOptions}
            placeholder="Select"
            value={newNotification.target}
          />
        </Col>
        <Col sm={dimensions && dimensions.width > 600 ? 3 : 12}>
          <span>Urgency</span>
          <Select
            className="app-Select"
            classNamePrefix="app-Select"
            onChange={event => onFieldChange({ target: { value: event } }, 'urgency')}
            options={urgencyOptions}
            placeholder="Select"
            value={newNotification.urgency}
          />
        </Col>
        <Col sm={dimensions && dimensions.width > 600 ? 3 : 12}>
          <span>Until</span>
          <NotificationDatePicker
            date={newNotification.until}
            onChange={onFieldChange}
          />
        </Col>
        {isEditing && (
          <Col sm={dimensions && dimensions.width > 600 ? 3 : 12}>
            <span>Active</span>
            <Select
              className="app-Select"
              classNamePrefix="app-Select"
              onChange={event => onFieldChange({ target: { value: event } }, 'active')}
              options={activeOptions}
              placeholder="Select"
              value={newNotification.active}
            />
          </Col>
        )}
      </Row>
      <Row className="action-row">
        <Col sm={12}>
          <button
            onClick={() => addElement('<a href="http://" target="_blank">Text</a>', activeLanguage)}
            type="button"
          >
            {'</a>'}
          </button>

          <button
            onClick={() => addElement('<b></b>', activeLanguage)}
            type="button"
          >
            {'<b>'}
          </button>
        </Col>
      </Row>
      <Row>
        <Col sm={12}>
          <div className="language-row">
            {languageButtons.map(lang => (
              <button
                className={classNames('language-button', {
                  'active': activeLanguage === lang,
                })}
                key={lang}
                onClick={() => setActiveLanguage(lang)}
                type="submit"
              >
                {lang.toUpperCase()}
              </button>
            ))}
          </div>
          <textarea
            onChange={event => onFieldChange(event, 'message', activeLanguage)}
            placeholder={`Notification message for ${activeLanguage.toUpperCase()}`}
            rows={5}
            value={(newNotification.message && newNotification.message[activeLanguage]) || ''}
          />
        </Col>
      </Row>
      <Row>
        <Col className="button-row" sm={12}>
          {!isEditing && <button onClick={addNew} type="submit">Add</button>}
        </Col>
      </Row>

    </div>
  );
};

CreateNotificationsForm.propTypes = {
  addElement: PropTypes.func.isRequired,
  addNew: PropTypes.func.isRequired,
  isEditing: PropTypes.bool,
  newNotification: PropTypes.object.isRequired,
  onFieldChange: PropTypes.func.isRequired,
};

export default CreateNotificationsForm;
