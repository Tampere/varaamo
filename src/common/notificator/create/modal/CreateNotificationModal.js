import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Modal } from 'react-bootstrap';

import CreateNotificationsForm from '../form/CreateNotificationsForm';
import DeleteConfirmationDialog from './DeleteConfirmationModal';

const CreateNotificationModal = (props) => {
  const [deleteModal, setDeleteModal] = useState(false);
  const {
    isOpen, onHide, selectedNotification, addElement, save, onDelete, onFieldChange,
  } = props;

  const deleteConfirm = () => {
    setDeleteModal(false);
    onDelete();
  };

  return (
    <React.Fragment>
      <Modal
        className="app-CreateNotificationModal"
        onHide={onHide}
        show={isOpen}
      >
        <Modal.Header>
          <Modal.Title>Edit notification</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <div className="modal-container">
            <CreateNotificationsForm
              addElement={addElement}
              addNew={save}
              isEditing
              newNotification={selectedNotification}
              onFieldChange={onFieldChange}
            />
          </div>
        </Modal.Body>
        <Modal.Footer className="modal-footer">
          <button onClick={() => setDeleteModal(true)} type="button">Delete</button>
          <button onClick={onHide} type="button">Close</button>
          <button onClick={save} type="button">Save</button>
        </Modal.Footer>
      </Modal>

      <DeleteConfirmationDialog
        deleteConfirm={deleteConfirm}
        isOpen={deleteModal}
        onHide={() => setDeleteModal(false)}
      />
    </React.Fragment>

  );
};

CreateNotificationModal.propTypes = {
  addElement: PropTypes.func.isRequired,
  save: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  onFieldChange: PropTypes.func.isRequired,
  onHide: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired,
  selectedNotification: PropTypes.object.isRequired,
};

export default CreateNotificationModal;
