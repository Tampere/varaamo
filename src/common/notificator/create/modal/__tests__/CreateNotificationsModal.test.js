import React from 'react';
import toJson from 'enzyme-to-json';
import simple from 'simple-mock';
import { shallow } from 'enzyme';

import CreateNotificationModal from '../CreateNotificationModal';
import CreateNotificationsForm from '../../form/CreateNotificationsForm';

describe('CreateNotificationsModal', () => {
  const defaultProps = {
    addElement: simple.stub(),
    save: simple.stub(),
    isOpen: false,
    onFieldChange: simple.stub(),
    onHide: simple.stub(),
    onDelete: simple.stub(),
    selectedNotification: {},
  };
  const getWrapper = props => shallow(<CreateNotificationModal {...defaultProps} {...props} />);

  test('Renders correcly', () => {
    const wrapper = getWrapper();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe('Form exists', () => {
    test('form exists', () => {
      const wrapper = getWrapper();
      const form = wrapper.find(CreateNotificationsForm);
      expect(form.length).toEqual(1);
    });
  });
});
