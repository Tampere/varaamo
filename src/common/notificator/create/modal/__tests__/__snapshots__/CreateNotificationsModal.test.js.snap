// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateNotificationsModal Renders correcly 1`] = `
<Fragment>
  <Modal
    animation={true}
    autoFocus={true}
    backdrop={true}
    bsClass="modal"
    className="app-CreateNotificationModal"
    dialogComponentClass={[Function]}
    enforceFocus={true}
    keyboard={true}
    manager={
      ModalManager {
        "add": [Function],
        "containers": Array [],
        "data": Array [],
        "handleContainerOverflow": true,
        "hideSiblingNodes": true,
        "isTopModal": [Function],
        "modals": Array [],
        "remove": [Function],
      }
    }
    onHide={[Function]}
    renderBackdrop={[Function]}
    restoreFocus={true}
    show={false}
  >
    <ModalHeader
      bsClass="modal-header"
      closeButton={false}
      closeLabel="Close"
    >
      <ModalTitle
        bsClass="modal-title"
        componentClass="h4"
      >
        Edit notification
      </ModalTitle>
    </ModalHeader>
    <ModalBody
      bsClass="modal-body"
      componentClass="div"
    >
      <div
        className="modal-container"
      >
        <CreateNotificationsForm
          addElement={[Function]}
          addNew={[Function]}
          isEditing={true}
          newNotification={Object {}}
          onFieldChange={[Function]}
        />
      </div>
    </ModalBody>
    <ModalFooter
      bsClass="modal-footer"
      className="modal-footer"
      componentClass="div"
    >
      <button
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
      <button
        onClick={[Function]}
        type="button"
      >
        Close
      </button>
      <button
        onClick={[Function]}
        type="button"
      >
        Save
      </button>
    </ModalFooter>
  </Modal>
  <DeleteConfirmationDialog
    deleteConfirm={[Function]}
    isOpen={false}
    onHide={[Function]}
  />
</Fragment>
`;
